### PR TITLE
Project Id Solution

### PR DESCRIPTION
**Pull request by thorrester**
### Short Summary
Updates project id logic to a more permanent solution

### Notes
- Moves project id logic out of `_RunManager` into a new class called `_ProjectRegistrar`
- Focus of new class is to (1) get project id or (2) create a new project given `name` and `repo`
- Getting project id and max project id have been moved to sql logic for performance
- Added new project/id and project/max_id routes 
- Added a new `ProjectQueryEngine` that has customized logic for the project registry
   - **NOTE** this is a slight departure since there is no longer a generic `QueryEngine` that all registries share
   - This is OK given that it will enable us to create registry specific logic on the backend if we ever need to without affecting public interfaces. There may be better ways to do this in the future though (don't know how good it is to call `_registry`. If we find that we are doing this a lot we should consider separating that logic out.

### CODE REVIEW

File: opsml/app/routes/projects.py [hunk 1]
```diff
@@ -64,3 +68,48 @@ async def project_metric_page(
...
+
+    return ProjectIdResponse(
+        project_id=registry.get_project_id(
+            project_name=project_name,
+            repository=repository,
+        )
+    )
+
+
+@router.get("/projects/max_id", response_model=ProjectIdResponse, name="project_id")
```
Comment (damonallison) [inline opsml/app/routes/projects.py:103]: what would you think about a POST `/projects` to create a new project if one does not already exist, which would return the ID as part of the response. This feels cleaner than having the client create a new ID.

File: opsml/projects/project.py [hunk 2]
```diff
@@ -2,21 +2,69 @@
...
+
+        logger.info("New project detected! Registering card")
+
+        card = ProjectCard(
+            name=self._project_info.name,
+            repository=self._project_info.repository,
+            contact=self._project_info.contact,
+            project_id=max_project + 1,
+        )
+        self.registries.project.register_card(card=card)
```
Comment (damonallison) [inline opsml/projects/project.py:63]: Or we could have the registration process create the ID as part of registration.

File: opsml/app/routes/projects.py (hunk 1 above)
Comment (thorrester) [reply opsml/app/routes/projects.py:103]: yea, good call. I think that sounds cleaner. I'll dig into it

File: opsml/registry/sql/base/client.py [hunk 3]
```diff
@@ -328,6 +329,52 @@ def registry_type(self) -> RegistryType:
...
+            params={
+                "project_name": project_name,
+                "repository": repository,
+            },
+        )
+
+        project_id = data.get("project_id")
+
+        assert isinstance(project_id, int)
+        return project_id
```
Comment (thorrester) [inline opsml/registry/sql/base/client.py:376]: Get project id as part of card registration

**Review by damonallison — APPROVED** [review]

**Pull request merged**